### DISCUSSION TITLE
Framing boxes (dropped shadow, rounded corners and more)

### DISCUSSION THREAD

**Discussion by Omikhleia** [sile-typesetter/sile#1313] [Show and tell]
Now working on a **framebox** package...

Think "a trimmed-down LaTeX fancybox package", in a way... 

The latter might have more options and features, maybe (I haven't checked for long), but my aim was just to provide the things I feel a need for, without extra considerations:
- dropped shadow
- rounded corners
- colors

The initial version is live in my repo and provides the following things:

![image](https://user-images.githubusercontent.com/18075640/154861227-737e97d9-74c1-4803-ad94-fd239f25c122.png)

Interestingly,maybe, for the curious: regarding the implementation, I just went for PDF graphic objects and used `SILE.outputter:drawSVG` (which has a slightly misleading name actually) to push them to the page. An interesting way to programmatically craft PDF graphics from code :rofl: ...

... I'm now considering some nicer framing that is a bit less obvious - so it might take some time, but I think the existing initial package is usable at least and worth a "show and tell". I'll update this message is the secondary aim is achieved!

**Comment by Omikhleia** [comment]
> I'm now considering some nicer framing that is a bit less obvious 

Oh well, it's going even better than I thought (although the current experimental code is kind of ugly and will clearly need a refactor ^^), so the idea here is to port (a decent subset) of [rough.js](https://github.com/rough-stuff/rough) to Lua so that to get this kind of framed boxes:

![image](https://user-images.githubusercontent.com/18075640/154873102-c0c5b223-9222-4546-9d4d-31626ad036dd.png)

![image](https://user-images.githubusercontent.com/18075640/154874570-a7a21afb-0d79-4f5e-b642-5494167e17ee.png)

Quick and dirty so far, but it does its trick IMHO. And that's something that might not exist yet even in (Lua)LaTeX, maybe? :laughing:

**Reply by alerque** [reply]
This is looking right spiffy. One suggestion might be to split up the port of Rough.js to Lua into a stand-alone LuaRocks library project and use it as a dependency for the SILE package instead of baking them together.

**Comment by Omikhleia** [comment]
Exposing option tuning, many fixes (... doh those arrays starting at 0 or 1 ^^) and "hachure" filling ported...

![image](https://user-images.githubusercontent.com/18075640/155312538-c4895877-8edf-42c6-8a11-55cb98b14337.png)

True, the (small subset of) roughjs part should be cleaned up, refactored and likely extracted at some point. Possibly extended too, for other needs (the current port just includes lines, rectangles and "hachure-fills")... Other features out of my mind that might wait a bit too (I'll be switching to another of my needs before tackling with them, but contributions would be welcome):
- Ability to select between `math.random` and a dedicated seeded pseudo-RNG (so that one can replicate the same shapes instead of getting new ones on re-runs)
- A "roughrules" package, for extending the "rules" package features with rough rules, underlining, etc. ^^

**Comment by Omikhleia** [comment]
A late update to this old "show and tell" topic: The features demonstrated here are available as part of the [ptable.sile](https://github.com/Omikhleia/ptable.sile) module.
